feat: add initional integration of rattler-build

nichmor opened the pull request:
WIP ⚠️ 🚧 

list of things to be tested/ implemented:
* selectors for new variants.yaml
* config_files naming should or should not contain channel_targets that is exposed by rattler-build recipe


<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [ ] Added a ``news`` entry

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

beckermr reviewed (changes requested):
Why are introducing a bunch of extra template files? I would assume we can use the ones we have already.

nichmor commented:
> Why are introducing a bunch of extra template files? I would assume we can use the ones we have already.

Good question! they will be deleted on final PR version - I'm using them in development process to clearly point what is needed new for rattler-build and will integrate it in the old templates

xhochy commented:
It would be nice if this were coordinated together with the existing `rattler-build` implementation in https://github.com/conda-forge/polarify-feedstock/tree/rattler-build

pavelzw commented:
The latest conda-smithy version that bela was working on is here: https://github.com/0xbe7a/conda-smithy/commit/1e818fc0ad4c7e5c9f3539ba9fa88cb70ed82707

nichmor commented:
Hey @beckermr! I would like to ask for your review :) 
I've added some notes in the PR description to explain what & how integration was done - hope that it will be easier to review with these notes.

jakirkham commented:
Feedback from the meeting earlier:

* Let's move the rattler-build schema into rattler-build
* Rename `conda_smithy.rattler_build` to something like `conda_smithy.rattler_build_compat` or similar
* Can we confirm that linter case covers other conda-forge lints (like duplicates in staged-recipes, bioconda checks, etc.)

Others should feel welcome to jump in and add/revise anything above

jakirkham commented:
May have missed this earlier, but with `conda smithy init` are we adding a flag to allow configuring whether `conda-build` or `rattler-build` is used? Or how is that handled?

Also there was a mention about a new `variant.yaml`. Do you have some more details about what that is and how it is being used?

Thanks for your help! 🙏

beckermr commented:
> Rename conda_smithy.rattler_build to something like conda_smithy.rattler_build_compat or similar

This is not what was mentioned on the call. The suggestion there as I understood it was to move that file into rattler build and rename it conda_compat or similar.

jakirkham commented:
Ah ok. Thanks for clarifying! 🙏

I may not have understood the suggestion

wolfv commented:
Currently rattler-build doesn't have any Python parts. I don't really see how it would be useful to move the rattler-build / conda-build integration anywhere else right now. If it's OK I would suggest to keep it in conda-smithy and we could revisit this when `rattler-build` has a Python API.

isuruf commented:
I meant a `rattler_build.conda_build_compat` module.  I didn't know that rattler-build didn't have a python API.
Still, I don't think we should be maintaining this code in conda-smithy. Maybe a small python package `rattler-conda-build-compat` would make sense here.

nichmor commented:
> I meant a `rattler_build.conda_build_compat` module. I didn't know that rattler-build didn't have a python API. Still, I don't think we should be maintaining this code in conda-smithy. Maybe a small python package `rattler-conda-build-compat` would make sense here.

hey @isuruf ! thanks a lot for your review!

I will extract the entire `rattler_build` package into a separate conda package that will be imported by conda-smithy and called directly `rattler-conda-build.render`, without the need of storing this code in `conda-smithy`.
Let me know if it's sounds good for you

isuruf commented:
Sounds great. Thanks for working on that.

nichmor commented:
> May have missed this earlier, but with `conda smithy init` are we adding a flag to allow configuring whether `conda-build` or `rattler-build` is used? Or how is that handled?
> 
> Also there was a mention about a new `variant.yaml`. Do you have some more details about what that is and how it is being used?
> 
> Thanks for your help! 🙏

hey @jakirkham ! thanks a lot for your input!

* during the execution of `conda smithy init` it will try to load first the `meta.yaml`. If it's not present, it switch to check if `recipe.yaml` can be located. If the `recipe.yaml` is present, during the creation of feedstock it writes to `conda-forge.yaml`
`conda_build_tool: rattler-build` so this later would tell `conda-smithy` ( for example on `rerender` step ) what tool to use for building or rendering.
* `variants.yaml` is a new [configuration file](https://prefix-dev.github.io/rattler-build/latest/variants/) used for rattler-build, similar how `conda_build_config.yaml` is for `conda-build`.  The "good news" is that `conda_build_config.yaml` is still supported ( if `variants.yaml` is missing ) by conda-smithy even when using rattler-build and it will generate correctly the CI configuration files. In [polarify-feedstock example](https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/recipe/recipe.yaml) we can see that `recipe.yaml `have `conda_build_config.yaml `and `channel_targets` are picked up correctly for `[azure configuration](https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/.ci_support/linux_64_channel_targetsconda-forge_polarify-rattler-build_dev.yaml)`

Please let me know if you have other questions / or suggestion regarding this.

thank you!

jakirkham commented:
Thanks for the info! 🙏

> during the execution of `conda smithy init` it will try to load first the `meta.yaml`. If it's not present, it switch to check if `recipe.yaml` can be located. If the `recipe.yaml` is present, during the creation of feedstock it writes to `conda-forge.yaml`
`conda_build_tool: rattler-build` so this later would tell `conda-smithy` ( for example on `rerender` step ) what tool to use for building or rendering.

Right I mean, can we please make this configurable at `conda smithy init` time?

Also, at least until we have more experience with it, the default of `conda smithy init` should be to use `conda-build`. Would like this to be an opt-in experience until we have more familiarity

> `variants.yaml` is a new [configuration file]( https://prefix-dev.github.io/rattler-build/latest/variants/ ) used for rattler-build, similar how `conda_build_config.yaml` is for `conda-build`.  The "good news" is that `conda_build_config.yaml` is still supported ( if `variants.yaml` is missing ) by conda-smithy even when using rattler-build and it will generate correctly the CI configuration files. In [polarify-feedstock example]( https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/recipe/recipe.yaml ) we can see that `recipe.yaml `have `conda_build_config.yaml `and `channel_targets` are picked up correctly for [azure configuration]( https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/.ci_support/linux_64_channel_targetsconda-forge_polarify-rattler-build_dev.yaml )

Am curious why we have two files that are doing the same thing. Would it be possible to just use `conda_build_config.yaml`?

Or alternatively if we want to move this to a new naming convention, could we start a CEP to motivate that change?

cc @jezdez (who may be interested in this as well)

nichmor commented:
> Thanks for the info! 🙏
> 
> > during the execution of `conda smithy init` it will try to load first the `meta.yaml`. If it's not present, it switch to check if `recipe.yaml` can be located. If the `recipe.yaml` is present, during the creation of feedstock it writes to `conda-forge.yaml`
> > `conda_build_tool: rattler-build` so this later would tell `conda-smithy` ( for example on `rerender` step ) what tool to use for building or rendering.
> 
> Right I mean, can we please make this configurable at `conda smithy init` time?
> 
> Also, at least until we have more experience with it, the default of `conda smithy init` should be to use `conda-build`. Would like this to be an opt-in experience until we have more familiarity
> 
> > `variants.yaml` is a new [configuration file](https://prefix-dev.github.io/rattler-build/latest/variants/) used for rattler-build, similar how `conda_build_config.yaml` is for `conda-build`.  The "good news" is that `conda_build_config.yaml` is still supported ( if `variants.yaml` is missing ) by conda-smithy even when using rattler-build and it will generate correctly the CI configuration files. In [polarify-feedstock example](https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/recipe/recipe.yaml) we can see that `recipe.yaml `have `conda_build_config.yaml `and `channel_targets` are picked up correctly for [azure configuration](https://github.com/nichmor/polarify-feedstock/blob/feat/conda-smithy-rattler-build/.ci_support/linux_64_channel_targetsconda-forge_polarify-rattler-build_dev.yaml)
> 
> Am curious why we have two files that are doing the same thing. Would it be possible to just use `conda_build_config.yaml`?
> 
> Or alternatively if we want to move this to a new naming convention, could we start a CEP to motivate that change?
> 
> cc @jezdez (who may be interested in this as well)

* I'm totally happy with adding a new option for `conda smithy init` but my only concern is that we would need to update also other services ( for example staged-recipes : https://github.com/conda-forge/staged-recipes/blob/main/.github/workflows/scripts/create_feedstocks.py#L242 ) so they could "understand" how to render a feedstock with new recipe.yaml which would "increase" functional complexity. On other hand , leverage to autodiscover what build tool to use in case of missing meta.yaml and presence of recipe.yaml seems more intuitive

* the new *variants.yaml* offer a new [selector syntax](https://prefix-dev.github.io/rattler-build/latest/selectors/#selectors-in-recipes) so both recipe.yaml + variants.yaml share same "grammar" and it is recommended ( and the only ) to configure `rattler-build`.
I've designed backward compatibility in `conda-smithy` of `conda_build_config.yaml` for `rattler-build` to support [this ](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml) configuration file so `out of the box` it's still can support recipe.yaml + conda_build_config.yaml ( which is not the recommended way because rattler-build itself does not support direct `conda_build_config.yaml` support ) 

Please let me know what you think about these points and how we can improve them if necessary

wolfv commented:
To add to what @nichmor said: rattler-build does not understand the 'comments-as-selectors' in an arbitrary conda_build_config.yaml. We haven't implemented that and are not really planning to implement it. Thankfully `conda-smithy` produces the rendered `conda_build_configs` and they are easy to parse and work with rattler-build.

isuruf commented on conda_smithy/lint_recipe.py:
```diff
@@ -1105,17 +1108,47 @@ def _format_validation_msg(error: "jsonschema.ValidationError"):
+    )
+
+
+def _lint_recipe_yaml(recipe_path):
+    with open(recipe_path) as fh:
+        meta = get_yaml().load(fh)
+
+    return validate_json_schema(
+        meta, schema_file=RATTLER_RECIPE_YAML_SCHEMA_FILE
+    )
```
The linter is supposed to enforce best practices. For eg: checking that the maintainers exist, tests are there, etc.
All this does is that the recipe follows the standard. This is less critical as rattler itself would be doing that.

nichmor replied:
done - implemented linting checks and extended test-suite for this

isuruf commented on tests/test_configure_feedstock.py:
```diff
@@ -578,37 +588,53 @@ def test_secrets(py_recipe, jinja_env):
+    expected_value = "1000"
+    conda_build_param = request.node.callspec.params["config_yaml"]
+    if conda_build_param == "rattler-build":
+        ci_support_filename = "linux_64_python3.8zlib1.2.12.yaml"
```
Why is this different?

isuruf commented on tests/test_configure_feedstock.py:
```diff
@@ -211,7 +211,13 @@ def test_py_matrix_on_azure(py_recipe, jinja_env):
+def test_stdlib_on_azure(stdlib_recipe, jinja_env, request):
+    conda_build_param = request.node.callspec.params["config_yaml"]
+    if conda_build_param == "rattler-build":
+        pytest.skip(
+            "skipping test for rattler-build usecase as we currently we don't have stdlib"
+        )
```
Is there a rattler-build issue we can link here?

nichmor replied:
yes, let me add it

isuruf commented on tests/test_configure_feedstock.py:
```diff
@@ -721,8 +755,13 @@ def test_migrator_compiler_version_recipe(
+        assert "win_64_c_compilervs2017python3.5.yaml" in rendered_variants
+    else:
+        assert "win_64_python3.8ruby2.7.yaml" in rendered_variants
+        assert "win_64_python3.10ruby3.2.yaml" in rendered_variants
```
Why is this different?

isuruf commented on conda_smithy/cli.py:
```diff
@@ -123,10 +131,21 @@ def __call__(self, args):
-        if args.recipe_directory:
+        # detect if it's old recipe or new one
+        meta: Union[MetaData, RattlerMetaData]
+        try:
```
Don't use a try statement. I would suggest one of the two.
1. Check that there's a `<recipe_dir>/recipe.yaml`
2. Check for `conda_build_tool` in `<recipe_dir>/conda-forge.yml`

nichmor replied:
thanks, will change it

nichmor replied:
Implemented this using has_rattler_recipe
https://github.com/conda-forge/conda-smithy/pull/1876/files#diff-af50d500653645ceb4a41789ee02aec35fa4c46b48c4635bb611ed88a5d3c3b9R141

let me know if it's ok

nichmor commented:
hey @isuruf I would like to ask for review again : I've implemented missing linting checks and extracted all rattler-build related code to rattler-build-conda-compat.

hey @jakirkham ! I wanted to let you know that I've implemented the conda-specific checks ( quote: `Can we confirm that linter case covers other conda-forge lints (like duplicates in staged-recipes, bioconda checks, etc.`) here: https://github.com/nichmor/rattler-build-conda-compat/blob/main/src/rattler_build_conda_compat/lint.py#L424
and moved the schema outside ( into rattler-build-conda-compat module ) . Also extended the current test suite for testing linting of recipe to support recipe.yaml testing.

isuruf commented on tests/test_configure_feedstock.py:
```diff
@@ -578,30 +592,42 @@ def test_secrets(py_recipe, jinja_env):
 
+    ci_support_filename = "linux_64_python2.7.yaml"
+    conda_build_param = request.node.callspec.params["config_yaml"]
+    if conda_build_param == "rattler-build":
+        ci_support_filename = "linux_64_python2.7zlib1000.yaml"
```
Why this change?

nichmor replied:
rattler-build expose all variants that are used during the rendering so because we generate filenames using them - for rattler-build they are little longer and contains all of them

isuruf replied:
Can we fix that in rattler-build or `ratller-build` or `rattler-build-conda-compat`. Otherwise this would include a lot of unnecessary information.

nichmor replied:
> Can we fix that in rattler-build or `ratller-build` or `rattler-build-conda-compat`. Otherwise this would include a lot of unnecessary information.

Yes - I can change it

nichmor replied:
done - it will generate the same name as conda-build does

nichmor replied:
until the new version of rattler-build-conda-compat will be picked up by solver the tests will fail temporary

nichmor commented:
hey @beckermr ! 
Additional templates were removed - so now we are extending existing ones.
Let me know what do you think
Thank you.

isuruf reviewed (changes requested):
Please rerender an existing feedstock that was rendered with latest released conda-smithy and make sure that there was no change.

nichmor commented:
> Please rerender an existing feedstock that was rendered with latest released conda-smithy and make sure that there was no change.

hey @isuruf ! This is the result:
https://github.com/nichmor/rattler-build-feedstock/pull/1
https://github.com/nichmor/py-rattler-feedstock/pull/1

there were some differences in amount of spaces in .sh scripts, so I've adjusted conda-smithy in this commit
https://github.com/conda-forge/conda-smithy/pull/1876/commits/780a11596b7c1313e08f9b32e5c262a878921d50

let me know if there is anything else I can do

nichmor commented:
hey @isuruf @beckermr ! I wanted to ask if there is anything else that need to be changed or improved.  Thank you!

wolfv commented on conda_smithy/lint_recipe.py:
```diff
@@ -1006,6 +1393,7 @@ def run_conda_forge_specific(meta, recipe_dir, lints, hints):
 
     # 5: Package-specific hints
     # (e.g. do not depend on matplotlib, only matplotlib-base)
+    # TODO: do the same for if selectors
```
should we take care of this? Is this about flattening the selectors?

isuruf replied:
I don't understand this comment

nichmor replied:
I've removed this comment. 
It was meant for `rattler-build-conda-compat package` to handle the situation when some of the packages are listed under `then/else ` branches. 
It is done here: https://github.com/nichmor/rattler-build-conda-compat/blob/main/src/rattler_build_conda_compat/loader.py#L90

wolfv reviewed (approved):
Looks great!

isuruf commented on conda_smithy/templates/run_osx_build.sh.tmpl:
```diff
@@ -93,17 +101,27 @@ if [[ "${BUILD_WITH_CONDA_DEBUG:-0}" == 1 ]]; then
 
+    {%- else %}
+
+    {{ BUILD_CMD }} --recipe ./{{ recipe_dir }} -m ./.ci_support/${CONFIG}.yaml --output-dir ${MINIFORGE_HOME}/conda-bld
```
`${EXTRA_CB_OPTIONS:-}` is not used here

nichmor replied:
right now `rattler-build` has a limited support of options ( for example, we don't have `--no-test` that is used in `EXTRA_CB_OPTIONS` ).

That's why I didn't used it here

nichmor replied:
some of the options that are used in `EXTRA_CB_OPTIONS` are not present yet in rattler-build 
( like `--output-id` ).

That's why I didn't used it

isuruf replied:
`--output-id` is added in debug mode which is correctly handled right now. `--no-test` is not supported?

nichmor replied:
you are right ! I've added EXTRA_CB_OPTIONS to rattler-build

isuruf commented on tests/test_configure_feedstock.py:
```diff
@@ -682,19 +701,24 @@ def test_migrator_downgrade_recipe(
 
+    conda_build_param = request.node.callspec.params["config_yaml"]
+    expected_value = "1000"
+
+    ci_support_filename = "linux_64_python2.7.yaml"
+
     with open(
         os.path.join(
             recipe_migration_cfep9_downgrade.recipe,
             ".ci_support",
-            "linux_64_python2.7.yaml",
+            ci_support_filename,
         )
     ) as fo:
         variant = yaml.safe_load(fo)
-        assert variant["zlib"] == ["1000"]
+        assert variant["zlib"] == [expected_value]
```
Can you reduce unnecessary diff like this?

nichmor replied:
yes, done.

beckermr reviewed (changes requested):
Please merge main into this pr as there will be conflicts and I'd like to see the final diff.

beckermr reviewed (changes requested):
conda-forge uses the `{{ stdlib('c') }}` jinja expression now. We'll need testing and support for that from rattler-build in order to merge this PR. Otherwise people won't be able to correctly use the compilers.

wolfv commented:
Hey @beckermr - yes, we'll add support for the `stdlib` jinja function to rattler-build. But shouldn't this be orthogonal to this PR? I think we would want to ratify the CEPs before allowing rattler-build on conda-forge anyways. Ideally we could still already merge this PR?

beckermr commented:
Thanks @wolfv! If we merge this PR, we are allowing support for rattler as soon as there is a release of conda-smithy. So, if we want to wait for the CEPS, which I agree we should, then we need to wait no matter what. So I think getting support in for the stdlib jinja and adding that to the tests should be able to happen before this is merged.

isuruf commented:
Maybe we can disallow publishing to `conda-forge main` label with rattler-build and create an issue with things to be done before that restriction is taken off?

wolfv commented:
I would very much prefer @isuruf suggestion so that we can move on from this big PR and iterate with a few smaller ones :)

beckermr commented:
Yes I agree with @isuruf. That will work great! I'll finish the rest of my review.

beckermr reviewed (changes requested):
I left a bunch of comments. This will need some refactoring and big changes before it can be merged.

wolfv commented on conda_smithy/lint_recipe.py:
```diff
@@ -1017,7 +1028,383 @@ def sort_osx(versions):
 
 
-def run_conda_forge_specific(meta, recipe_dir, lints, hints):
+def lintify_recipe_yaml(
```
@beckermr is that then opposing isuruf's previous comments about making an external package for `rattler-build-conda-compat`? Because this was on specific request. I think we'd be fine for moving the functionality back into smithy.

beckermr replied:
That request pertained only to the APIs in conda-build, not to the stuff smithy handles. The request as I understood it was to mirror the split in apis between conda-build and conda-smithy.

nichmor replied:
hey @beckermr ! 
I will move linting checks back into conda-smithy as I agree with you that it will give more control for conda-forge team.
the schema lint will be in rattler_build_conda_compat

nichmor replied:
@beckermr 
I've moved all rattler lint checks in `conda_smithy/rattler_linter.py`. Let me know please what do you think about this approach. Thanks

wolfv commented:
@beckermr is the "big change" you are talking about the linting move from `rattler-build-conda-compat` back into smithy? Or did you mean something else (that seems doable).

Thanks for the review!

beckermr commented:
yes @wolfv - I meant the linting stuff in the packages.

nichmor commented on conda_smithy/lint_recipe.py:
```diff
@@ -1017,7 +1028,383 @@ def sort_osx(versions):
 
 
-def run_conda_forge_specific(meta, recipe_dir, lints, hints):
+def lintify_recipe_yaml(
```
Agree.
I didn't want to make this PR bigger and increase it's complexity with additional refactor of `lint_recipe.py` but it's a good candidate for future improvement.

nichmor commented on tests/conftest.py:
```diff
@@ -105,14 +105,23 @@ def config_yaml(testing_workdir, recipe_dirname):
+    if config_yaml_param_value == "rattler-build":
+        recipe_name = "recipe.yaml"
+    else:
+        recipe_name = "meta.yaml"
```
Agree. Changed the fixture a little so it will `yield` a `namedtuple`

beckermr commented on conda_smithy/rattler_linter.py:
```diff
@@ -0,0 +1,590 @@
+        return """Recipe name has invalid characters. only lowercase alpha, numeric, underscores, hyphens and dots allowed"""
+
+
+def run_conda_forge_specific(
```
The duplication in this function is especially egregious. Let's fix that before merging as it is a few line refactor I think.

nichmor replied:
done - we now pass rattler_lint to differentiate how to extract necessary fields
let me know wdyt

beckermr replied:
thank you!

isuruf commented on conda_smithy/lint_recipe.py:
```diff
@@ -1017,15 +1029,393 @@ def sort_osx(versions):
-    sources_section = get_section(meta, "source", lints)
-    requirements_section = get_section(meta, "requirements", lints)
-    outputs_section = get_section(meta, "outputs", lints)
+    if not rattler_lint:
+        # Retrieve sections from meta.yaml
+        package_section = get_section(meta, "package", lints)
+        extra_section = get_section(meta, "extra", lints)
+        sources_section = get_section(meta, "source", lints)
+        requirements_section = get_section(meta, "requirements", lints)
+        outputs_section = get_section(meta, "outputs", lints)
+    else:
+        # Retrieve sections from recipe.yaml
+        package_section = meta.get("package", {})
+        extra_section = meta.get("extra", {})
+        sources_section = meta.get("source", [])
+        requirements_section = rattler_loader.load_all_requirements(meta)
+        outputs_section = meta.get("outputs", {})
```
Can we move this logic to `get_section`?

nichmor replied:
to be honest I think it's better to keep them explicitly in this way
If we move it to `get_section`, we will add also there `rattler_lint` and just use meta.get in case of rattler-build.
keeping them in `run_conda_forge` function seems a more transparent option, without touching other functions

wdyt?

isuruf replied:
I'd rather that we have these in helper functions so that the main logic stays clean. Thoughts?

nichmor replied:
Agree. 
I've moved them into get_rattler_section and get_meta_section which are under get_section umbrella.

wdyt?

beckermr reviewed (approved):
There is still work left to do to reduce duplication in the linter, but lgtm!

nichmor commented:
hey @isuruf ! I've made some refactor changes to reuse common code across meta and recipe.yaml linting as you suggested

nichmor commented:
hey @isuruf ! Let me know if there is anything else you want me to refactor
Thank you!

isuruf commented on conda_smithy/lint_recipe.py:
```diff
@@ -1081,15 +1156,221 @@ def sort_osx(versions):
+            meta, recipe_dir, lints, hints, rattler_lint=True
+        )
+
+    # 15: Check if we are using legacy patterns
+    rattler_linter.lint_legacy_patterns(requirements_section)
```
Why is this different from conda-build?

nichmor commented:
As discussed with @isuruf, I've removed the linter from this PR. It will be added in the next one with proper refactoring.

isuruf reviewed (commented):
Looks good. Tests still fail though

nichmor commented:
> Looks good. Tests still fail though

fixed. rattler-build was cached in the CI on older version

isuruf commented:
Looks good. Can you add a news entry?

nichmor commented:
> Looks good. Can you add a news entry?

Added

jaimergp commented:
A piece was missing. Added in https://github.com/conda-forge/conda-smithy/pull/1977.
